### PR TITLE
UI app shell

### PR DESCRIPTION
**Pull request by meirish**
After adding gzip to the static assets I looked at what else we could do to increase perceived load time for the application and this is a small thing we can do.

This uses a fork of `ember-app-shell` to scrape the HTML from a specified route in the build process and overwrite the ember app's `index.html` file. So a user will see the "splash page" immediately when the HTML loads, not after the JS app boots and begins navigating.

Note: I've opened a PR to upstream the changes to `ember-app-shell` - uncertain how that will pan out, but if it does get merged we should move off o the fork.

Note 2: puppeteer will install Chromium when it gets installed so now anyone building the UI will get an extra download of 170-280Mb or so (see the note here: https://github.com/GoogleChrome/puppeteer#installation ). If we should do something about this or change defaults or add the `PUPPETEER_SKIP_CHROMIUM_DOWNLOAD` in dev, let me know.

### CODE REVIEW

**Comment by joshuaogle** [comment]
Is there a trick to getting this to show in the app? By the time I see the logo I also see the footer (and if I'm reading this right it should be the logo by itself, right?). Looks good from what I can tell, just wanted to make sure I'm seeing the right thing

**Comment by meirish** [comment]
Nope, you’re seeing the right thing! The difference is if you view source, there’s HTML there now not just script tags. That html is the same as our loading page so you see the html, then it gets replaced by the app code that will exactly match it. If you navigate to /ui/app-shell it’ll stay on that page.

**Comment by joshuaogle** [comment]
Cool, I was trying to see if that's what it was doing, but it was still super fast with the network slowed. It does feel a little snappier 👍

**Comment by meirish** [comment]
OK, so this _doesn't_ work for Enterprise currently (lots of jumpiness on the loading screen) - will work on fixing that and comment on it later.

File: ui/app/routes/app-shell.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import Ember from 'ember';
+
+export default Ember.Route.extend({
+  version: Ember.inject.service(),
+  beforeModel() {
+    return this.get('version').fetchVersion();
+  },
+
+  afterModel() {
```
Comment (thrashr888) [inline ui/app/routes/app-shell.js]: Did you mean to leave this in? If so it could use a comment as-to why.

**Comment by meirish** [comment]
Some terrible gifs (I didn't spend the time to sync the beginning) but I think you can still see the difference. These are on full speed network with a local vault server so it's the best case scenario. The after one you see the Vault logo almost immediately, then Vault Enterprise,  the before has just a white page and then Vault Enterprise.

## Before

loading-before

## After

loading-after

**Comment by meirish** [comment]
Going to close this - I think in the end the puppeteer install is too much and I may open a PR doing this by hand which will be a much lighter touch.

**Pull request closed without merging**